subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/AttributeMatchingVariantSelector.java

```diff
@@ -115,24 +118,60 @@ private ResolvedArtifactSet doSelect(ResolvedVariantSet producer) {
         throw new NoMatchingVariantSelectionException(producer.asDescribable().getDisplayName(), componentRequested, producer.getVariants(), matcher);
     }
 
-    private List<Pair<ResolvedVariant, ConsumerVariantMatchResult.ConsumerVariant>> tryDisambiguate(List<Pair<ResolvedVariant, ConsumerVariantMatchResult.ConsumerVariant>> candidates) {
+    private List<Pair<ResolvedVariant, ConsumerVariantMatchResult.ConsumerVariant>> tryDisambiguate(AttributeMatcher matcher, List<Pair<ResolvedVariant, ConsumerVariantMatchResult.ConsumerVariant>> candidates) {
+        if (candidates.size() == 2) {
+            // Short circuit logic when only 2 candidates
+            return compareCandidates(matcher, candidates.get(0), candidates.get(1))
+                .map(Collections::singletonList)
+                .orElse(candidates);
+        }
+
         List<Pair<ResolvedVariant, ConsumerVariantMatchResult.ConsumerVariant>> shortestTransforms = Lists.newArrayListWithExpectedSize(candidates.size());
+        candidates.sort(Comparator.comparingInt(candidate -> candidate.right.depth));
+
+        // Need to remember if a further element was matched by an earlier one, no need to consider it then
+        boolean[] hasBetterMatch = new boolean[candidates.size()];
 
-        for (Pair<ResolvedVariant, ConsumerVariantMatchResult.ConsumerVariant> candidate : candidates) {
-            boolean newCandidate = true;
-            for (int i = 0; i < shortestTransforms.size(); i++) {
-                Pair<ResolvedVariant, ConsumerVariantMatchResult.ConsumerVariant> shorterTransform = shortestTransforms.get(i);
-                if (shorterTransform.right.transformation.endsWith(candidate.right.transformation)) {
-                    newCandidate = false;
-                    shortestTransforms.set(i, candidate);
-                } else if (candidate.right.transformation.endsWith(shorterTransform.right.transformation)) {
-                    newCandidate = false;
+        for (int i = 0; i < candidates.size(); i++) {
+            if (hasBetterMatch[i]) {
```

Comment:
:x: Correct me if I'm wrong but I think this test is always false, since `i` is always incremented, and that `j>i`.

Reply:
No, let's say you have three candidates. The first candidate is better than the second, so below we will set `hasBetterMatch[1]` to true. When we finish comparing the first candidate with the others, we move on to the second. And immediately know we can skip it.